### ISSUE TITLE
Lack of homogeneity in R2L outputs

### ISSUE THREAD

**Issue by ngeiswei** [opencog/opencog#2193]
See for instance

```
;; r2l output corresponding to (chat "Bob lives in China")

   (SetLink
      (InheritanceLink
         (ConceptNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
         (ConceptNode "in" (stv 0.032258064 0.0012484394))
      )
      (ImplicationLink
         (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
         (PredicateNode "bob" (stv 0.16666667 0.0012484394))
      )
      (InheritanceLink
         (SatisfyingSetLink
            (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
         )
         (ConceptNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
      )
      (InheritanceLink
         (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
         (ConceptNode "China" (stv 0.032258064 0.0012484394))
      )
      (ImplicationLink
         (PredicateNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
         (PredicateNode "in" (stv 0.16666667 0.0012484394))
      )
      (EvaluationLink
         (PredicateNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
         (ListLink
            (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
            (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
         )
      )
      (EvaluationLink
         (DefinedLinguisticPredicateNode "definite")
         (ListLink
            (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
         )
      )
      (EvaluationLink
         (PredicateNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
         (ListLink
            (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
         )
      )
      (InheritanceLink
         (InterpretationNode "sentence@0c2ab58f-7150-4259-8cd2-52ddf260d7f1_parse_0_interpretation_$X")
         (DefinedLinguisticConceptNode "ImperativeSpeechAct")
      )
      (InheritanceLink
         (SpecificEntityNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
         (DefinedLinguisticConceptNode "female" (stv 0.050000001 0.0012484394))
      )
      (InheritanceLink
         (SpecificEntityNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
         (ConceptNode "China" (stv 0.032258064 0.0012484394))
      )
      (InheritanceLink
         (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
         (DefinedLinguisticConceptNode "imperative")
      )
   )
```

Notice the implication

```
      (ImplicationLink
         (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
         (PredicateNode "bob" (stv 0.16666667 0.0012484394))
      )
```

But then see

```
;; r2l output corresponding to (chat "Bob likes chicken feet")

   (SetLink
      (InheritanceLink
         (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
         (ConceptNode "Bob" (stv 0.025641026 0.0012484394))
      )
      (EvaluationLink
         (DefinedLinguisticPredicateNode "definite")
         (ListLink
            (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
         )
      )
      (ImplicationLink
         (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
         (PredicateNode "like" (stv 0.125 0.0012484394))
      )
      (InheritanceLink
         (ConceptNode "feet@68a4daa2-1f53-4c0f-8443-806099311b49")
         (ConceptNode "foot" (stv 0.025641026 0.0012484394))
      )
      (EvaluationLink
         (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
         (ListLink
            (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
            (ConceptNode "feet@68a4daa2-1f53-4c0f-8443-806099311b49")
         )
      )
      (EvaluationLink
         (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
         (ListLink
            (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
         )
      )
      (InheritanceLink
         (InterpretationNode "sentence@e943fc5b-cf61-4ed1-a172-ab09d545ae67_parse_0_interpretation_$X")
         (DefinedLinguisticConceptNode "DeclarativeSpeechAct")
      )
      (InheritanceLink
         (ConceptNode "chicken@43d36c82-1eb0-4e42-b694-68e21023b5fa")
         (ConceptNode "chicken" (stv 0.025641026 0.0012484394))
      )
      (InheritanceLink
         (ConceptNode "feet@68a4daa2-1f53-4c0f-8443-806099311b49")
         (ConceptNode "chicken@43d36c82-1eb0-4e42-b694-68e21023b5fa")
      )
      (InheritanceLink
         (SpecificEntityNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
         (DefinedLinguisticConceptNode "male")
      )
      (InheritanceLink
         (SpecificEntityNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
         (ConceptNode "Bob" (stv 0.025641026 0.0012484394))
      )
      (InheritanceLink
         (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
         (DefinedLinguisticConceptNode "present")
      )
   )
```

Now Bob's instance relates to Bob by an inheritance link

```
      (InheritanceLink
         (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
         (ConceptNode "Bob" (stv 0.025641026 0.0012484394))
      )
```

Not a complete contradiction but really completely homogeneous.

**Comment by ngeiswei** [comment]
Hmm, rules like `tense`, `adverbialpp`, `advmod` consider these [nous like Bob] as verbs. Don't really understand that...

**Comment by bgoertzel** [comment]
Somehow it seems the sentence "Bob lives in China" is being interpreted as
an imperative with "Bob" as a verb (meaning to bounce up and down a bit),
i.e.

"I command you to bob people's lives up and down, in the country of China
!!!"

This is a meaningful parse, but should not be the top-ranked one...

-- Ben

On Tue, Jun 14, 2016 at 8:19 PM, ngeiswei <notifications@github.com> wrote:

> See for instance
>
> ;; r2l output corresponding to (chat "Bob lives in China")
>
>    (SetLink
>       (InheritanceLink
>          (ConceptNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
>          (ConceptNode "in" (stv 0.032258064 0.0012484394))
>       )
>       (ImplicationLink
>          (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
>          (PredicateNode "bob" (stv 0.16666667 0.0012484394))
>       )
>       (InheritanceLink
>          (SatisfyingSetLink
>             (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
>          )
>          (ConceptNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
>       )
>       (InheritanceLink
>          (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
>          (ConceptNode "China" (stv 0.032258064 0.0012484394))
>       )
>       (ImplicationLink
>          (PredicateNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
>          (PredicateNode "in" (stv 0.16666667 0.0012484394))
>       )
>       (EvaluationLink
>          (PredicateNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
>          (ListLink
>             (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
>             (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
>          )
>       )
>       (EvaluationLink
>          (DefinedLinguisticPredicateNode "definite")
>          (ListLink
>             (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
>          )
>       )
>       (EvaluationLink
>          (PredicateNode "in@b8dbff0f-ffe4-4775-9905-f9c87f29f63d")
>          (ListLink
>             (ConceptNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
>          )
>       )
>       (InheritanceLink
>          (InterpretationNode "sentence@0c2ab58f-7150-4259-8cd2-52ddf260d7f1_parse_0_interpretation_$X")
>          (DefinedLinguisticConceptNode "ImperativeSpeechAct")
>       )
>       (InheritanceLink
>          (SpecificEntityNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
>          (DefinedLinguisticConceptNode "female" (stv 0.050000001 0.0012484394))
>       )
>       (InheritanceLink
>          (SpecificEntityNode "China@b5d7ee9c-4cb1-4a70-b7ea-4dec2db35558")
>          (ConceptNode "China" (stv 0.032258064 0.0012484394))
>       )
>       (InheritanceLink
>          (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
>          (DefinedLinguisticConceptNode "imperative")
>       )
>    )
>
> Notice the implication
>
>       (ImplicationLink
>          (PredicateNode "bob@04c3177d-bb4d-4a44-ae30-f4acbf1ea59f")
>          (PredicateNode "bob" (stv 0.16666667 0.0012484394))
>       )
>
> But then see
>
> ;; r2l output corresponding to (chat "Bob likes chicken feet")
>
>    (SetLink
>       (InheritanceLink
>          (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>          (ConceptNode "Bob" (stv 0.025641026 0.0012484394))
>       )
>       (EvaluationLink
>          (DefinedLinguisticPredicateNode "definite")
>          (ListLink
>             (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>          )
>       )
>       (ImplicationLink
>          (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
>          (PredicateNode "like" (stv 0.125 0.0012484394))
>       )
>       (InheritanceLink
>          (ConceptNode "feet@68a4daa2-1f53-4c0f-8443-806099311b49")
>          (ConceptNode "foot" (stv 0.025641026 0.0012484394))
>       )
>       (EvaluationLink
>          (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
>          (ListLink
>             (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>             (ConceptNode "feet@68a4daa2-1f53-4c0f-8443-806099311b49")
>          )
>       )
>       (EvaluationLink
>          (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
>          (ListLink
>             (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>          )
>       )
>       (InheritanceLink
>          (InterpretationNode "sentence@e943fc5b-cf61-4ed1-a172-ab09d545ae67_parse_0_interpretation_$X")
>          (DefinedLinguisticConceptNode "DeclarativeSpeechAct")
>       )
>       (InheritanceLink
>          (ConceptNode "chicken@43d36c82-1eb0-4e42-b694-68e21023b5fa")
>          (ConceptNode "chicken" (stv 0.025641026 0.0012484394))
>       )
>       (InheritanceLink
>          (ConceptNode "feet@68a4daa2-1f53-4c0f-8443-806099311b49")
>          (ConceptNode "chicken@43d36c82-1eb0-4e42-b694-68e21023b5fa")
>       )
>       (InheritanceLink
>          (SpecificEntityNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>          (DefinedLinguisticConceptNode "male")
>       )
>       (InheritanceLink
>          (SpecificEntityNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>          (ConceptNode "Bob" (stv 0.025641026 0.0012484394))
>       )
>       (InheritanceLink
>          (PredicateNode "likes@dc51995f-6e69-4d8e-91a7-0d5992dba6c6")
>          (DefinedLinguisticConceptNode "present")
>       )
>    )
>
> Now Bob's instance relates to Bob by an inheritance link
>
>       (InheritanceLink
>          (ConceptNode "Bob@c0bce0fd-d30f-402e-8897-259fa248473d")
>          (ConceptNode "Bob" (stv 0.025641026 0.0012484394))
>       )
>
> Not a complete contradiction but really homogeneous.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencog/opencog/issues/2193>, or mute the thread
> <https://github.com/notifications/unsubscribe/AFolXFcUC1nNTxdg7uhoF_xXF8-V9pNfks5qLpxFgaJpZM4I1Qlu>
> .
>



-- 
Ben Goertzel, PhD
http://goertzel.org

Super-benevolent super-intelligence is the thought the Global Brain is
currently struggling to form...

**Comment by linas** [comment]
1) The issue goes away if you say "Robert lives in China".  There are a handful of common names for which such ambiguities occur. ("May" is one - Mae, Mabel, "march" is another.)

2) Its not the top-ranked one: it's ranking is identical to the correct parse.  The R2L pipeline ignores the costs that LG uses to indicate ranking. It randomly picks one, and goes with it.

3) This can be fixed in LG.  The LG dicts already contain a cost that causes "Tom" the given name to be preferred over "Tom" the tomato, the tom-tom drum, and the tom-verb (to hit a drum).

4) The correct long-term solution is to use the LG weights (costs) as "advisories", and have R2L and PLN adjust those costs as more information becomes available.

**Comment by linas** [comment]
FYI, Point 3) is fixed in commit opencog/link-grammar/commit/6780d33c45653551f582600bb57272416b4dbae6 

Updating and renewing the corpus-statistics based weighting, per issue opencog/link-grammar#293 ''might'', maybe, avoid some future parse-ranking issues. Maybe.  Its on rather ambiguous theoretical footing right now, and is not immediately or obviously better, and that cannot be cured without a big investment of time.  Continuing with langauge learning would be a better option.

**Comment by ngeiswei** [comment]
Using Robert fixes it (something I really didn't expect), thanks!

**Comment by ngeiswei** [comment]
They are other inconsistencies with the adverb "in", sometimes predicate, sometimes concept, should an adverb be a predicate or a concept?

**Comment by linas** [comment]
"in" is not really an adverb, its a preposition.   In English, there are a fixed number of prepositions, (you can write them all down)  and they are used to express relationships e.g. spatial (next to, inside of, ...) temporal (after, before) political (belongs to) .  Linguistically, they are somewhat verb-like, in that they can be used in an huge variety of ways, being very highly flexible, and having rather few rules that constrain their use.

Mathematically, because the express relationships that can be true or false, it seems like they should be predicates.  For example: "Paris is the capital of France" -- "Paris is the capital of Germany" -- "Paris is in Germany", "Paris belongs to the French", "This book belongs to Ben", etc.  

Think of it this way: "belongs" is some verb, "to" is the preposition, and the combination "belongs to" is the predicate.  That is what should be happening. Don't know if R2L actually does this correctly or not.

**Comment by linas** [comment]
Viz:  X "belongs to" Y, X "is in" Y, X "is next to" Y,  and in general X R Y where R is the relation, a predicate, built up from a verb-preposition pair.

**Comment by ngeiswei** [comment]
Is "in" really always a preprosition? Like in the sentence "I moved in", is "in" a preposition?

Anyway, OK, so in the sentence "Robert lives in Hong Kong", "in" really is a preposition. So I suppose the advmod rule should not be triggered then, right?

BTW, should an adverb be a concept or a predicate?

**Comment by linas** [comment]
For "I moved in", then "in" is an adverb, not a preposition. 

For "Robert lives in Hong Kong", "in" really is a preposition, and probably the advmod rule should not trigger.  If it is triggering, it's probably due to a bug in relex: link-grammar itself is using the J link, which always likes to prepositional objects, so there is no doubt or uncertainty there.

For expressions like "I moved in" ... I really don't know. How could PLN do something useful with that? What could it deduce that makes sense?  There's a kind-of missing word that you have to deduce from the context:  "I moved in [closer, the get a better view]" or "I moved in[to the apartment last week]"

http://www.thefreedictionary.com/in -- even when grammatically an "adverb", its still behaving like it should really be forming a predicate, along with the verb.  I don't quite see how it can be free-standing, live by itself, from the point of view of PLN, from the point of view of deducing something logical.

**Comment by linas** [comment]
This is kind of an example of where the higher reaches of relex+R2L start falling apart, without a lot of careful tinkering.  So -- "I moved in[to the apartment]" still has the general form X R Y where X==I and R==relation=="move into" and Y has to be deduced from the context of the conversation.  That is, Y is a blank, a place-holder, a variable that is not known, and reasoning has to be applied to determine what Y might be.  It is a form of "reference resolution".

This is exactly where reasoning starts becoming very interesting and important: first, to realize that there is an implied Y, and second, to deduce what Y is -- or at least, await for future evidence that might reveal what Y could be -- Imagine that "I moved in" is the first sentence of a short story.

**Comment by ngeiswei** [comment]
Oh, according to the gender r2l rule China is a female

```
      (InheritanceLink
         (SpecificEntityNode "China@a6d3b47a-3a03-4ecf-9cc2-18032c2bc858")
         (DefinedLinguisticConceptNode "female" (stv 0.055555556 0.0012484394))
      )
```

**Comment by linas** [comment]
Heh. Welcome to the fuzzy world of NLP. Link-grammar has a list of male and female names, and "China" is a not-uncommon female name. It also has a list of locations/place-names, and China also appears there as well. There is no way to tell, from syntactic analysis, which is intended, and so *both* are reported, (each in a different parse) with an equal score. RelEx converts these tags into gender markup (a different parse, with the same score, should not have the gender tag in it).

You have multiple choices:
A) use logic and reasoning to determine if "China" is a person or a place, and change the truth value on that parse to reflect this new information.
B) prevent Relex from reporting gender, and then, later on if reasoning determines that China is a woman, add a gender tag. -- but if reasoning determines its a location, add a location tag.

The general idea is that, for each parse, find additional information via, logic, deduction, reasoning, to determine if that parse is more or less correct, and update the truth value as needed.

In the medium term, option B might be better -- in the long term, however, it is impossible to parse without also knowing semantics.  Many sentences are not ambiguous, but many more are.  Parsing is mostly "high quality guessing", reasoning has to provide the missing details.

**Comment by anitzkin** [comment]
Well, actually . . . the spatial meanings of prepositions in English are
considered their foundational meanings, but they have extended,
metaphorical meanings, in countless domains, often abstract, such as "I
work in Anthropology" or "I'm in a tizzy" or about a zillion others.  Their
use is entirely constrained by rules.

Regarding your "belongs" example, you probably want to retain the
distinction between phrasal verbs, such as "belong to" and verbs plus
ancillary prepostiions, because they obey different grammatical rules.  A
phrasal verb consisting of more than one word.  To see the difference, you
might consider "I looked up the meaning of the word" (phrasal verb 'look
up') versus "I looked up at the sky."  (not a phrasal verb).  The
difference is that "up the meaning of the word" is not a phrase; the
constituents are "look up" and "the meaning of the word." Versus the second
example, where the consitutents are "look" and "up at the sky" which is
clearly a phrase.

Of course all of this analysis depends on considering two things which you
have repeatedly told me need not be part of your grammar -- semantics and
phrase structure.  So, suit yourself.

**Comment by anitzkin** [comment]
I went through a lot of trouble to design the advmod-rule so that it could
support reasoniong.  Please, try to understand what it's doing before you
muck about it with it.

Prepositions are treated as adverbs now in R2L when they introduce
"adverbials" which are phrases, playing the roles of adverbs.  For example
"Bill moved into a new apartment" -- "into a new apartment" is essentially
an adverb.  It modifies "moved"; it tells you how / where Bill moved.
Since we don't work explicitly with phrases, I needed to attach the
adverby-ness of the phrase to its head -- the preposition.  Which I think
works very well.   This is an important modification.  Because it makes no
sense at all to treat "into" as the object of the verb, which is what you
were doing before.

**Comment by linas** [comment]
Hi Aaron, 

Its a data representation issue. Relations are always predicates that have truth values:  X R Y is true or false, X "moved into" Y is a predicate whose truth value can be ascertained.

By contrast,  "Bill moved" is also a kind of predicate, which can be true or false.  However, ascertaining that truth is harder, without involving the modifier:  "Bill moved" can be true, but "Bill moved into the apartment" can be false, because maybe Bill moved into the elevator.   The point is that reasoning becomes trickier -- detaching the adverbial phrase has the effect of "erasing" an important piece of knowledge, or at least, putting it where it is harder to reach and manipulate.

I dunno. It all depends on what one is trying to do.   The "we don't work with explicitly phrases" seems like another jab in the dependency vs phrase structure war, and, from where I sit, its got nothing to do with anything that's important.  The point is that all sorts of clusters of words interact with each other and carry meaning.   Both PSG and DG are two different ways of highlighting some of the most important inter-relationships, but there are certainly many many more, of secondary importance.  There's no "one right way", all the words are interconnected.  

The classic example is "I heard the church bells ring on Sunday"  where the important relations are "heard-ring-bell" implying sound (as opposed to other meanings for "ring") and "bell-church-Sunday" implying a physical, architectural building (as opposed to the abstract concept of "church" e.g. "the Vatican")  Neither PSG or DG tell you about either of these relations.  In PSG, the words end up in different phrases; in DG, there is no direct linkage between them, because, in neither case is the meaning conveyed directly through syntactic means.

**Comment by linas** [comment]
closing, because the original bug report had to do with a link-grammar parse-liklihood error, confusing "bob" the verb with "Bob" the given name.  The error is fixed in link-grammar release 5.3.8
